File: tracers/php/AlgorithmVisualizer.php
<?php

/**
 * Thư viện tracer cho PHP.
 *
 * Phải sinh ra command list GIỐNG HỆT bản JavaScript tham chiếu — PLAN.md Phụ lục C.
 * Chỉ dùng thư viện chuẩn: sandbox của Piston không có mạng.
 *
 * Ký tự escape trong chuỗi kết quả dựng từ mã số (`chr(92)`) chứ không viết literal:
 * chuỗi PHP có hai lớp escape và rất dễ sai một lớp mà không ai thấy.
 */

final class Av
{
    /** Tiền tố tách kênh lệnh khỏi stdout thường — §3.5 quy tắc 1. */
    public const PREFIX = "\x1e@AV|";

    /** Giới hạn số nguyên an toàn — §3.5 quy tắc 2b. int của PHP là 64 bit, của JS là 53. */
    public const SAFE_INTEGER_LIMIT = 9007199254740991;

    private static int $nextKey = 0;

    public static function newKey(): string
    {
        $key = 'k' . self::$nextKey;
        self::$nextKey++;
        return $key;
    }

    public static function record(?string $key, string $method, array $args): void
    {
        $parts = [];
        foreach ($args as $arg) {
            $parts[] = self::dump($arg);
        }
        echo self::PREFIX
            . '{"key":' . ($key === null ? 'null' : self::quote($key))
            . ',"method":' . self::quote($method)
            . ',"args":[' . implode(',', $parts) . ']}'
            . PHP_EOL;
    }

    /**
     * Tự tuần tự hóa thay vì json_encode: json_encode in số thực nguyên là "1.0" trong
     * một số cấu hình và escape "/" thành gạch chéo có escape. Khác một ký tự là bộ
     * tuân thủ đỏ mà không rõ nguyên nhân.
     */
    public static function dump(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_int($value)) {
            if (abs($value) > self::SAFE_INTEGER_LIMIT) {
                throw new RuntimeException(
                    "Gia tri $value vuot khoang so nguyen an toan cua giao thuc. "
                    . 'JavaScript se lam tron gia tri nay.'
                );
            }
            return (string) $value;
        }
        if (is_float($value)) {
            return self::dumpFloat($value);
        }
        if (is_string($value)) {
            return self::quote($value);
        }
        if (is_array($value)) {
            return array_is_list($value) ? self::dumpList($value) : self::dumpMap($value);
        }
        if ($value instanceof AvNode) {
            return self::quote($value->key);
        }
        return 'null';
    }

    private static function dumpList(array $items): string
    {
        $parts = [];
        foreach ($items as $item) {
            $parts[] = self::dump($item);
        }
        return '[' . implode(',', $parts) . ']';
    }

    private static function dumpMap(array $map): string
    {
        $parts = [];
        foreach ($map as $key => $item) {
            $parts[] = self::quote((string) $key) . ':' . self::dump($item);
        }
        return '{' . implode(',', $parts) . '}';
    }

    /** NaN và vô cực không phải JSON hợp lệ — §3.5 quy tắc 3. */
    private static function dumpFloat(float $value): string
    {
        if (is_nan($value)) {
            return '{' . self::quote('$num') . ':' . self::quote('NaN') . '}';
        }
        if (is_infinite($value)) {
            return '{' . self::quote('$num') . ':'
                . self::quote($value > 0 ? 'Infinity' : '-Infinity') . '}';
        }
        if ($value == floor($value) && abs($value) < 1e15) {
            return (string) (int) $value;
        }
        return rtrim(rtrim(sprintf('%.17g', $value), '0'), '.');
    }

    public static function quote(string $text): string
    {
        $bs = chr(92);
        $out = '"';
        $length = strlen($text);
        for ($i = 0; $i < $length; $i++) {
            $c = $text[$i];
            $code = ord($c);
            if ($c === '"') {
                $out .= $bs . '"';
            } elseif ($c === $bs) {
                $out .= $bs . $bs;
            } elseif ($code === 10) {
                $out .= $bs . 'n';
            } elseif ($code === 13) {
                $out .= $bs . 'r';
            } elseif ($code === 9) {
                $out .= $bs . 't';
            } elseif ($code < 32) {
                $out .= $bs . 'u' . sprintf('%04x', $code);
            } else {
                $out .= $c;
            }
        }
        return $out . '"';
    }
}

/** Gốc chung: mọi tracer chỉ là một khoá cộng danh sách phương thức chuyển tiếp. */
abstract class AvNode
{
    public string $key;

    /** Phương thức được phép chuyển tiếp thẳng sang command list. */
    protected const FORWARD = [];

    public function __construct(string $className, array $args)
    {
        $this->key = Av::newKey();
        Av::record($this->key, $className, $args);
    }

    public function __call(string $name, array $args): void
    {
        if (!in_array($name, static::FORWARD, true)) {
            throw new BadMethodCallException(
                static::class . " khong co phuong thuc $name"
            );
        }
        Av::record($this->key, $name, $args);
    }

    public function destroy(): void
    {
        Av::record($this->key, 'destroy', []);
    }

    public function reset(): void
    {
        Av::record($this->key, 'reset', []);
    }

    protected static function titleArgs(?string $title): array
    {
        return $title === null ? [] : [$title];
    }
}

class Array2DTracer extends AvNode
{
    protected const FORWARD = [
        'set', 'patch', 'depatch', 'select', 'selectRow', 'selectCol',
        'deselect', 'deselectRow', 'deselectCol',
    ];

    public function __construct(?string $title = null, string $className = 'Array2DTracer')
    {
        parent::__construct($className, self::titleArgs($title));
    }
}

class Array1DTracer extends AvNode
{
    protected const FORWARD = ['set', 'patch', 'depatch', 'select', 'deselect'];

    public function __construct(?string $title = null, string $className = 'Array1DTracer')
    {
        parent::__construct($className, self::titleArgs($title));
    }

    public function chart(?AvNode $tracer): void
    {
        Av::record($this->key, 'chart', [$tracer?->key]);
    }
}

class ChartTracer extends Array1DTracer
{
    public function __construct(?string $title = null)
    {
        parent::__construct($title, 'ChartTracer');
    }
}

class ScatterTracer extends Array2DTracer
{
    public function __construct(?string $title = null)
    {
        parent::__construct($title, 'ScatterTracer');
    }
}

class LogTracer extends AvNode
{
    protected const FORWARD = ['set', 'print', 'println', 'printf'];

    public function __construct(?string $title = null)
    {
        parent::__construct('LogTracer', self::titleArgs($title));
    }
}

class MarkdownTracer extends AvNode
{
    protected const FORWARD = ['set'];

    public function __construct(?string $title = null)
    {
        parent::__construct('MarkdownTracer', self::titleArgs($title));
    }
}

class GraphTracer extends AvNode
{
    protected const FORWARD = [
        'set', 'directed', 'weighted', 'addNode', 'updateNode', 'removeNode',
        'addEdge', 'updateEdge', 'removeEdge', 'layoutCircle', 'layoutTree',
        'layoutRandom', 'visit', 'leave', 'select', 'deselect',
    ];

    public function __construct(?string $title = null)
    {
        parent::__construct('GraphTracer', self::titleArgs($title));
    }

    public function log(?AvNode $tracer): void
    {
        Av::record($this->key, 'log', [$tracer?->key]);
    }
}

abstract class AvLayoutNode extends AvNode
{
    /** @param AvNode[] $children */
    public function __construct(string $className, array $children)
    {
        parent::__construct($className, [array_map(fn(AvNode $c) => $c->key, $children)]);
    }

    public function add(AvNode $child, ?int $index = null): void
    {
        Av::record($this->key, 'add', $index === null ? [$child->key] : [$child->key, $index]);
    }

    public function remove(AvNode $child): void
    {
        Av::record($this->key, 'remove', [$child->key]);
    }

    public function removeAll(): void
    {
        Av::record($this->key, 'removeAll', []);
    }
}

class VerticalLayout extends AvLayoutNode
{
    public function __construct(array $children)
    {
        parent::__construct('VerticalLayout', $children);
    }
}

class HorizontalLayout extends AvLayoutNode
{
    public function __construct(array $children)
    {
        parent::__construct('HorizontalLayout', $children);
    }
}

final class Layout
{
    public static function setRoot(AvNode $node): void
    {
        Av::record(null, 'setRoot', [$node->key]);
    }
}

final class Tracer
{
    /**
     * Cắt một khung hình. Không truyền số dòng thì lấy từ ngăn xếp của người gọi —
     * `debug_backtrace` cho số dòng chính xác nên không phải bù offset như JavaScript.
     */
    public static function delay(?int $lineNumber = null): void
    {
        if ($lineNumber === null) {
            $frames = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
            $lineNumber = $frames[0]['line'] ?? 0;
        }
        Av::record(null, 'delay', [$lineNumber]);
    }
}

File: tracers/ruby/algorithm_visualizer.rb
# Thư viện tracer cho Ruby.
#
# Phải sinh ra command list GIỐNG HỆT bản JavaScript tham chiếu — PLAN.md Phụ lục C.
# Chỉ dùng thư viện chuẩn: sandbox của Piston không có mạng.

module AlgorithmVisualizer
  # Tiền tố tách kênh lệnh khỏi stdout thường — §3.5 quy tắc 1.
  PREFIX = 30.chr + '@AV|'

  # Giới hạn số nguyên an toàn — §3.5 quy tắc 2b. Integer của Ruby không giới hạn,
  # của JavaScript thì có: không chặn ở đây thì hai ngôn ngữ ra command list khác nhau.
  SAFE_INTEGER_LIMIT = 9_007_199_254_740_991

  @next_key = 0

  class << self
    def new_key
      key = "k#{@next_key}"
      @next_key += 1
      key
    end

    def record(key, method, args)
      payload = { 'key' => key, 'method' => method, 'args' => args.map { |a| normalize(a) } }
      $stdout.puts(PREFIX + dump(payload))
    end

    def normalize(value)
      case value
      when Integer
        if value.abs > SAFE_INTEGER_LIMIT
          raise "Gia tri #{value} vuot khoang so nguyen an toan cua giao thuc. " \
                'JavaScript se lam tron gia tri nay.'
        end
        value
      when Float
        # NaN và vô cực không phải JSON hợp lệ — §3.5 quy tắc 3
        return { '$num' => 'NaN' } if value.nan?
        return { '$num' => value.positive? ? 'Infinity' : '-Infinity' } if value.infinite?

        value
      when Array then value.map { |item| normalize(item) }
      when Hash then value.to_h { |k, v| [k.to_s, normalize(v)] }
      else value
      end
    end

    # Tự tuần tự hóa thay vì dùng `json`: bản JSON của Ruby in số thực nguyên là "1.0"
    # còn JavaScript in "1". Khác một ký tự là bộ tuân thủ đỏ.
    def dump(value)
      case value
      when nil then 'null'
      when true, false then value.to_s
      when Integer then value.to_s
      when Float then dump_float(value)
      when String then quote(value)
      when Symbol then quote(value.to_s)
      when Array then '[' + value.map { |item| dump(item) }.join(',') + ']'
      when Hash then '{' + value.map { |k, v| "#{quote(k.to_s)}:#{dump(v)}" }.join(',') + '}'
      else quote(value.to_s)
      end
    end

    def dump_float(value)
      return value.to_i.to_s if value.finite? && value == value.to_i && value.abs < 1e15

      value.to_s
    end

    def quote(text)
      out = +'"'
      text.each_char do |c|
        code = c.ord
        out << case c
               when '"' then 92.chr + '"'
               when 92.chr then 92.chr + 92.chr
               else
                 if code < 32
                   escape_control(code)
                 else
                   c
                 end
               end
      end
      out << '"'
    end

    def escape_control(code)
      case code
      when 10 then 92.chr + 'n'
      when 13 then 92.chr + 'r'
      when 9 then 92.chr + 't'
      else format('%su%04x', 92.chr, code)
      end
    end
  end

  # Gốc chung: mọi tracer chỉ là một khoá cộng danh sách phương thức chuyển tiếp.
  class Node
    attr_reader :key

    def initialize(class_name, args)
      @key = AlgorithmVisualizer.new_key
      AlgorithmVisualizer.record(@key, class_name, args)
    end

    def destroy
      AlgorithmVisualizer.record(@key, 'destroy', [])
    end

    def reset
      AlgorithmVisualizer.record(@key, 'reset', [])
    end

    def self.forward(*names)
      names.each do |name|
        define_method(name) do |*args|
          AlgorithmVisualizer.record(@key, name.to_s, args)
        end
      end
    end
  end

  def self.title_args(title)
    title.nil? ? [] : [title]
  end

  class Array2DTracer < Node
    def initialize(title = nil, class_name = 'Array2DTracer')
      super(class_name, AlgorithmVisualizer.title_args(title))
    end

    forward :set, :patch, :depatch, :select, :selectRow, :selectCol,
            :deselect, :deselectRow, :deselectCol
  end

  class Array1DTracer < Node
    def initialize(title = nil, class_name = 'Array1DTracer')
      super(class_name, AlgorithmVisualizer.title_args(title))
    end

    forward :set, :patch, :depatch, :select, :deselect

    def chart(tracer)
      AlgorithmVisualizer.record(@key, 'chart', [tracer&.key])
    end
  end

  class ChartTracer < Array1DTracer
    def initialize(title = nil)
      super(title, 'ChartTracer')
    end
  end

  class ScatterTracer < Array2DTracer
    def initialize(title = nil)
      super(title, 'ScatterTracer')
    end
  end

  class LogTracer < Node
    def initialize(title = nil)
      super('LogTracer', AlgorithmVisualizer.title_args(title))
    end

    forward :set, :print, :println, :printf
  end

  class MarkdownTracer < Node
    def initialize(title = nil)
      super('MarkdownTracer', AlgorithmVisualizer.title_args(title))
    end

    forward :set
  end

  class GraphTracer < Node
    def initialize(title = nil)
      super('GraphTracer', AlgorithmVisualizer.title_args(title))
    end

    forward :set, :directed, :weighted, :addNode, :updateNode, :removeNode,
            :addEdge, :updateEdge, :removeEdge, :layoutCircle, :layoutTree,
            :layoutRandom, :visit, :leave, :select, :deselect

    def log(tracer)
      AlgorithmVisualizer.record(@key, 'log', [tracer&.key])
    end
  end

  class LayoutNode < Node
    def initialize(class_name, children)
      super(class_name, [children.map(&:key)])
    end

    def add(child, index = nil)
      AlgorithmVisualizer.record(@key, 'add', index.nil? ? [child.key] : [child.key, index])
    end

    def remove(child)
      AlgorithmVisualizer.record(@key, 'remove', [child.key])
    end

    def removeAll
      AlgorithmVisualizer.record(@key, 'removeAll', [])
    end
  end

  class VerticalLayout < LayoutNode
    def initialize(children)
      super('VerticalLayout', children)
    end
  end

  class HorizontalLayout < LayoutNode
    def initialize(children)
      super('HorizontalLayout', children)
    end
  end

  module Layout
    def self.setRoot(node)
      AlgorithmVisualizer.record(nil, 'setRoot', [node.key])
    end
  end

  module Tracer
    # Cắt một khung hình. Không truyền số dòng thì lấy từ ngăn xếp của người gọi —
    # `caller_locations` cho số dòng chính xác nên không phải bù offset như JavaScript.
    def self.delay(line_number = nil)
      line_number ||= caller_locations(1, 1)&.first&.lineno || 0
      AlgorithmVisualizer.record(nil, 'delay', [line_number])
    end
  end
end

include AlgorithmVisualizer
